query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java

```diff
@@ -77,18 +81,25 @@ public Void apply(EmbeddedCacheManager embeddedCacheManager) {
          postIndex(indexUpdater, progressState, notifier);
       } else {
          DataConversion keyDataConversion = cache.getKeyDataConversion();
-         Set<Class<?>> classSet = new HashSet<>();
+         Set<Class<?>> classSet = new HashSet<>(keys.size());
+         List<CompletableFuture<Object>> updates = new ArrayList<>(keys.size());
```

Comment:
Sorry to ping pong back and forth. The preferred way of doing this is with https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/util/concurrent/CompletionStages.java#L44

This way you don't need to create another collection and the stages are as minimal as possible.